crates/exex/exex/src/wal/cache.rs

```diff
@@ -1,61 +1,70 @@
 use std::collections::{BTreeMap, VecDeque};
 
+use dashmap::DashMap;
 use reth_exex_types::ExExNotification;
-use reth_primitives::BlockNumHash;
+use reth_primitives::{BlockNumHash, B256};
 
-/// The block cache of the WAL. Acts as a mapping of `File ID -> List of Blocks`.
-///
-/// For each notification written to the WAL, there will be an entry per block written to
-/// the cache with the same file ID. I.e. for each notification, there may be multiple blocks in the
-/// cache.
+/// The block cache of the WAL.
 ///
 /// This cache is needed to avoid walking the WAL directory every time we want to find a
-/// notification corresponding to a block.
+/// notification corresponding to a block or a block corresponding to a hash.
 #[derive(Debug)]
-pub struct BlockCache(BTreeMap<u64, VecDeque<CachedBlock>>);
+pub struct BlockCache {
+    /// A mapping of `File ID -> List of Blocks`.
+    ///
+    /// For each notification written to the WAL, there will be an entry per block written to
+    /// the cache with the same file ID. I.e. for each notification, there may be multiple blocks
+    /// in the cache.
+    files: BTreeMap<u64, VecDeque<CachedBlock>>,
+    /// A mapping of `Block Hash -> Block`.
+    ///
+    /// For each [`ExExNotification::ChainCommitted`] notification, there will be an entry per
+    /// block.
+    blocks: DashMap<B256, CachedBlock>,
```

Comment:
looks like we only access this for insert rn.
DashMap seems fine here, alternatively a rwlock would also work

Reply:
yes, and we will later access it when reading the WAL from `ExExNotificationsWIthHead`.

yeah, an `RwLock` would be easier, so we can have the whole `BlockCache` behind it, and not just one field inside

Reply:
hmm actually no, we will have only `blocks` field in the future so it's fine to have a dashmap here https://github.com/paradigmxyz/reth/pull/11266#discussion_r1777654333